docs/design-documents/hal/0003-explicit-pinmap-extension.md

```diff
@@ -0,0 +1,474 @@
+```
+
+Explicit pinmap:
+
+```
+SPI(PinName mosi, PinName miso, PinName sclk, PinName ssel = NC);
+SPI(PinName mosi, PinName miso, PinName sclk, PinName ssel, use_gpio_ssel_t);
+
+SPI(const spi_pinmap_t &explicit_pinmap);
+SPI(const spi_pinmap_t &explicit_pinmap, PinName ssel);
+
+```
+
+Note that for SPI class currently, we have 2 constructors. One is used when `ssel` is to be driven by hardware and the other one if `ssel` is controlled manually.
+In the new version, the second constructor is a bit different. Does not provide `use_gpio_ssel_t`, but has `PinName ssel` instead. This is because in the case when we want to manually handle the `ssel` pin, then `ssel` pin in the given `explicit_pinmap` will be undefined (`NC`) and `ssel` pin must be provided separately. In this case we don't need additional `use_gpio_ssel_t` any more.
+
+### `constexpr` utility functions
+
+C++ 14 provides `constexpr` specifier which can be used to create utility functions to lookup the pinmap tables in compile-time and create the pin-mapping without adding pinmap tables to the image.
+
+Example pinmap lookup function below:
+
+```
+constexpr PinMap get_pwm_pinmap(const PinName pin)
+{
+    for (const PinMap &pinmap : PINMAP_PWM) {
+        if (pinmap.pin == pin) {
+            return {pin, pinmap.peripheral, pinmap.function};
+        }
+    }
+    return {NC, NC, NC};
+}
+```
+
+For targets which do not support explicit pinmap the utility functions will populate the pinmap structure only with given pins (backward compatibility):
+
+```
+constexpr spi_pinmap_t get_spi_pinmap(const PinName mosi, const PinName miso, const PinName sclk, const PinName ssel)
+{
+    return {NC, mosi, NC, miso, NC, sclk, NC, ssel, NC};
+}
+```
+
+C++ 14 is not available on ARM C5, so for backward compatibility when ARM C5 is used the utility functions will be defined as follows: `const PinMap get_pwm_pinmap(const PinName pin)`. Using `const` instead `constexpr` will add pinmap tables to the image (no ROM savings).
+
+### Example usage
+
+```
+int main()
+{
+    /* Regular use (master) */
+    SPI spi(D1, D2, D3, D4);
+
+    /* Explicit pinmap */
+    const spi_pinmap_t explicit_spi_pinmap = {SPI_1, D1, 2, D2, 2, D3, 2, D4, 2};
+    SPI spi(explicit_spi_pinmap);
+
+    /* Explicit pinmap with constexpr */
+    constexpr spi_pinmap_t explicit_spi_pinmap = get_spi_pinmap(D1, D2, D3, D4);
+    SPI spi(explicit_spi_pinmap);
+
+    return 0;
+}
+```
+
+Note that the following use of the pinmap utility function: `SPI spi(get_spi_pinmap_t(D1, D2, D3, D4));` on ARM compiler leads to unexpected behavior. In the mentioned case, the pinmap table will be added to the image (no ROM savings). Details below:
+
+![ARM constexpr issue](img/arm_constexpr_issue.png)
+
+It is strongly suggested that the result of `get_xxx_pinmap()` calls should normally be used in explicitly constant expression contexts like initializing a constexpr variable, to guarantee that the tables are not pulled in.
```

Comment:
Can you provide an example how to do it vs should not do (code snippet) ?

Reply:
Actually the code snippets are provided, but maybe again it is confusing. 
I will add that the suggested form is the third form from the examples. 
BTW this will be probably changed and the `SPI spi(get_spi_pinmap_t(D1, D2, D3, D4));`  form will lead to the compiler error.

Reply:
fixed